### PR TITLE
[zoe] Add new port

### PR DESCRIPTION
**Pull request by winsoft666**
If this PR adds a new port, please uncomment and fill out this checklist:

- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md)
- [x] The name of the port matches an existing name for this component on https://repology.org/ if possible, and/or is strongly associated with that component on search engines.
- [x] Optional dependencies are resolved in exactly one way. For example, if the component is built with CMake, all `find_package` calls are REQUIRED, are satisfied by `vcpkg.json`'s declared dependencies, or disabled with [CMAKE_DISABLE_FIND_PACKAGE_Xxx](https://cmake.org/cmake/help/latest/variable/CMAKE_DISABLE_FIND_PACKAGE_PackageName.html)
- [x] The versioning scheme in `vcpkg.json` matches what upstream says.
- [x] The license declaration in `vcpkg.json` matches what upstream says.
- [x] The installed as the "copyright" file matches what upstream says.
- [x] The source code of the component installed comes from an authoritative source.
- [x] The generated "usage text" is accurate. See [adding-usage](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/examples/adding-usage.md) for context.
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is in the new port's versions file.
- [x] Only one version is added to each modified port's versions file.

### CODE REVIEW

**Comment by FrankXie05** [comment]
```
/mnt/vcpkg-ci/buildtrees/zoe/src/41c3c2d161-9b4d7598fd.clean/src/entry_handler.cpp: In member function ‘zoe::Result zoe::EntryHandler::_asyncTaskProcess()’:
/mnt/vcpkg-ci/buildtrees/zoe/src/41c3c2d161-9b4d7598fd.clean/src/entry_handler.cpp:363:25: error: ‘sleep_for’ is not a member of ‘std::this_thread’
  363 |       std::this_thread::sleep_for(std::chrono::milliseconds(100));
      |                         ^~~~~~~~~
/mnt/vcpkg-ci/buildtrees/zoe/src/41c3c2d161-9b4d7598fd.clean/src/entry_handler.cpp:373:27: error: ‘sleep_for’ is not a member of ‘std::this_thread’
  373 |         std::this_thread::sleep_for(std::chrono::milliseconds(50));
      |                           ^~~~~~~~~
```
Maybe need include `<thread.h>`  or wrong namespace.
[install-x64-linux-dbg-out.log](https://github.com/microsoft/vcpkg/files/12844111/install-x64-linux-dbg-out.2.log)

**Comment by winsoft666** [comment]
> ```
> /mnt/vcpkg-ci/buildtrees/zoe/src/41c3c2d161-9b4d7598fd.clean/src/entry_handler.cpp: In member function ‘zoe::Result zoe::EntryHandler::_asyncTaskProcess()’:
> /mnt/vcpkg-ci/buildtrees/zoe/src/41c3c2d161-9b4d7598fd.clean/src/entry_handler.cpp:363:25: error: ‘sleep_for’ is not a member of ‘std::this_thread’
>   363 |       std::this_thread::sleep_for(std::chrono::milliseconds(100));
>       |                         ^~~~~~~~~
> /mnt/vcpkg-ci/buildtrees/zoe/src/41c3c2d161-9b4d7598fd.clean/src/entry_handler.cpp:373:27: error: ‘sleep_for’ is not a member of ‘std::this_thread’
>   373 |         std::this_thread::sleep_for(std::chrono::milliseconds(50));
>       |                           ^~~~~~~~~
> ```
> 
> Maybe need include `<thread.h>` or wrong namespace. [install-x64-linux-dbg-out.log](https://github.com/microsoft/vcpkg/files/12844111/install-x64-linux-dbg-out.2.log)

I have fixed this issue, could you try again?

**Comment by FrankXie05** [comment]
> > ```
> > /mnt/vcpkg-ci/buildtrees/zoe/src/41c3c2d161-9b4d7598fd.clean/src/entry_handler.cpp: In member function ‘zoe::Result zoe::EntryHandler::_asyncTaskProcess()’:
> > /mnt/vcpkg-ci/buildtrees/zoe/src/41c3c2d161-9b4d7598fd.clean/src/entry_handler.cpp:363:25: error: ‘sleep_for’ is not a member of ‘std::this_thread’
> >   363 |       std::this_thread::sleep_for(std::chrono::milliseconds(100));
> >       |                         ^~~~~~~~~
> > /mnt/vcpkg-ci/buildtrees/zoe/src/41c3c2d161-9b4d7598fd.clean/src/entry_handler.cpp:373:27: error: ‘sleep_for’ is not a member of ‘std::this_thread’
> >   373 |         std::this_thread::sleep_for(std::chrono::milliseconds(50));
> >       |                           ^~~~~~~~~
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Maybe need include `<thread.h>` or wrong namespace. [install-x64-linux-dbg-out.log](https://github.com/microsoft/vcpkg/files/12844111/install-x64-linux-dbg-out.2.log)
> 
> I have fixed this issue, could you try again?

This is error from CI check.

File: ports/zoe/vcpkg.json [hunk 1]
```diff
@@ -0,0 +1,23 @@
+{
+  "name": "zoe",
+  "version": "3.0",
+  "port-version": 1,
```
Comment (FrankXie05) [inline ports/zoe/vcpkg.json:4]:
```suggestion
```
The `+1` operation of `port-verison` will only be performed when the port is modified.
Simply, one PR gets `+1` at a time.

File: versions/z-/zoe.json [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    {
+      "git-tree": "b1c5c5bb8d55ed4a3b7a763dc19c69666694e6c4",
+      "version": "3.2",
+      "port-version": 0
+    },
+    {
+      "git-tree": "bbac35ef9a8843e0f75c2c8823c386909686fa50",
+      "version": "3.1",
+      "port-version": 0
+    },
```
Comment (FrankXie05) [inline versions/z-/zoe.json:17]:
```suggestion
  "versions": [
    {
      "git-tree": "24d50f143b93df2ab26fc6420ef6648cac18005a",
      "version": "3.0",
      "port-version": 1
    },
    {
      "git-tree": "b1c5c5bb8d55ed4a3b7a763dc19c69666694e6c4",
      "version": "3.2",
      "port-version": 0
    },
    {
      "git-tree": "bbac35ef9a8843e0f75c2c8823c386909686fa50",
      "version": "3.1",
      "port-version": 0
    },
```
Just run command `./vcpkg x-add-version zoe --overwrite-version`

**Review by FrankXie05 — CHANGES_REQUESTED** [review]

**Comment by FrankXie05** [comment]
CI error: https://dev.azure.com/vcpkg/public/_build/results?buildId=95324&view=artifacts&pathAsName=false&type=publishedArtifacts

```
FAILED: src/CMakeFiles/zoe-static.dir/file_util.cpp.o 
/android-ndk-r25c/toolchains/llvm/prebuilt/linux-x86_64/bin/clang++ --target=x86_64-none-linux-android21 --sysroot=/android-ndk-r25c/toolchains/llvm/prebuilt/linux-x86_64/sysroot -DNOMINMAX -DUNICODE -DWITH_OPENSSL -DZOE_STATIC -D_UNICODE -I/mnt/vcpkg-ci/buildtrees/zoe/src/b4e3a319e2-1e96496453.clean/./include -isystem /mnt/vcpkg-ci/installed/x64-android/include -g -DANDROID -fdata-sections -ffunction-sections -funwind-tables -fstack-protector-strong -no-canonical-prefixes -D_FORTIFY_SOURCE=2 -Wformat -Werror=format-security -frtti -fexceptions  -fPIC   -fno-limit-debug-info    -std=gnu++11 -fPIC -pthread -MD -MT src/CMakeFiles/zoe-static.dir/file_util.cpp.o -MF src/CMakeFiles/zoe-static.dir/file_util.cpp.o.d -o src/CMakeFiles/zoe-static.dir/file_util.cpp.o -c /mnt/vcpkg-ci/buildtrees/zoe/src/b4e3a319e2-1e96496453.clean/src/file_util.cpp
/mnt/vcpkg-ci/buildtrees/zoe/src/b4e3a319e2-1e96496453.clean/src/file_util.cpp:50:3: error: use of undeclared identifier 'fseeko64'
  fseeko64(f, 0L, SEEK_END);
  ^
/mnt/vcpkg-ci/buildtrees/zoe/src/b4e3a319e2-1e96496453.clean/src/file_util.cpp:51:19: error: use of undeclared identifier 'ftello64'; did you mean 'ftello'?
  int64_t fsize = ftello64(f);
                  ^~~~~~~~
                  ftello
/android-ndk-r25c/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/include/stdio.h:251:7: note: 'ftello' declared here
off_t ftello(FILE* __fp);
      ^
/mnt/vcpkg-ci/buildtrees/zoe/src/b4e3a319e2-1e96496453.clean/src/file_util.cpp:185:12: error: use of undeclared identifier 'fseeko64'
    return fseeko64(f, offset, origin);
           ^
3 errors generated.
```
[install-x64-android-dbg-out.log](https://github.com/microsoft/vcpkg/files/12865993/install-x64-android-dbg-out.log)

File: ports/zoe/portfile.cmake [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+)
+
+string(COMPARE EQUAL "${VCPKG_LIBRARY_LINKAGE}" "static" LIBRARY_LINKAGE_HAS_STATIC)
+string(COMPARE EQUAL "${VCPKG_CRT_LINKAGE}" "static" ZOE_USE_STATIC_CRT)
+
+if(${LIBRARY_LINKAGE_HAS_STATIC})
+	set(ZOE_BUILD_SHARED_LIBS OFF)
+else()
+	set(ZOE_BUILD_SHARED_LIBS ON)
+endif()
```
Comment (dg0yt) [inline ports/zoe/portfile.cmake:16]:
```suggestion
string(COMPARE EQUAL "${VCPKG_LIBRARY_LINKAGE}" "dynamic" ZOE_BUILD_SHARED_LIBS)
string(COMPARE EQUAL "${VCPKG_CRT_LINKAGE}" "static" ZOE_USE_STATIC_CRT)
```

BTW vcpkg already set `BUILD_SHARED_LIBS`, the standard variable.

File: ports/zoe/portfile.cmake [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+        -DZOE_USE_STATIC_CRT:BOOL=${ZOE_USE_STATIC_CRT}
+        -DZOE_BUILD_TESTS:BOOL=OFF
+)
+
+vcpkg_cmake_install()
+
+if(EXISTS "${CURRENT_PACKAGES_DIR}/lib/cmake/zoe")
+    vcpkg_cmake_config_fixup(CONFIG_PATH lib/cmake/zoe)
+elseif(EXISTS "${CURRENT_PACKAGES_DIR}/share/libGet")
+    vcpkg_cmake_config_fixup(CONFIG_PATH share/zoe)
```
Comment (dg0yt) [inline ports/zoe/portfile.cmake:31]: This will detect but not touch files in `/share/libGet`.

**Comment by winsoft666** [comment]
I successfully tested the following command on Windows 11, but automatic checks were not successful.
```
vcpkg install crashrpt-lite:x64-windows
```

**Comment by winsoft666** [comment]
> @FrankXie05 I add a new port that named crashrpt-lite on this pr. I don't know if this operation is compliant.

I have removed crashrpt-lite port.

**Comment by winsoft666** [comment]
@FrankXie05 @FrankXie05 Could you help me merge this pr?

**Review by FrankXie05 — DISMISSED** [review]

**Comment by winsoft666** [comment]
> @winsoft666 Only one question: Is the port `Header-only`? :)

No

**Comment by FrankXie05** [comment]
> > @winsoft666 Only one question: Is the port `Header-only`? :)
> 
> No

Then we should provide usage for this port to use lib. :)

**Comment by winsoft666** [comment]
> > > @winsoft666 Only one question: Is the port `Header-only`? :)
> > 
> > 
> > No
> 
> Then we should provide usage for this port to use lib. :)

Do you need me to do anything?

**Comment by FrankXie05** [comment]
@winsoft666 Yes, we need export cmake config files.

**Comment by winsoft666** [comment]
> @winsoft666 Yes, we need export cmake config files.

Just add usage file into port, like these?
```txt usage
The package zoe is compatible with built-in CMake targets:

    find_package(zoe REQUIRED)
    target_link_libraries(main PRIVATE zoe)
```

**Comment by winsoft666** [comment]
@FrankXie05 Thanks for your PR, I have merged.

**Comment by FrankXie05** [comment]
The usage has been tested successfully locally.
```
zoe provides CMake targets:

  # this is heuristically generated, and may not be correct
  find_package(zoe CONFIG REQUIRED)
  target_link_libraries(main PRIVATE zoe::zoe)
```

File: ports/zoe/portfile.cmake [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+        -DZOE_BUILD_TESTS:BOOL=OFF
+)
+
+vcpkg_cmake_install()
+
+if(EXISTS "${CURRENT_PACKAGES_DIR}/lib/cmake/zoe")
+    vcpkg_cmake_config_fixup(CONFIG_PATH lib/cmake/zoe)
+elseif(EXISTS "${CURRENT_PACKAGES_DIR}/share/zoe")
+    vcpkg_cmake_config_fixup(CONFIG_PATH share/zoe)
+endif()
```
Comment (dg0yt) [inline ports/zoe/portfile.cmake:26]: Why doesn't upstream always install to the same location... Both paths are good for all platforms.
And this code supports the no-config case (due to lack of unconditional  `else`). Do we want this?

Comment (FrankXie05) [inline ports/zoe/portfile.cmake:26]: Just add this.
```suggestion
vcpkg_cmake_config_fixup(CONFIG_PATH share/zoe)
```

**Review by FrankXie05 — CHANGES_REQUESTED** [review]

File: ports/zoe/portfile.cmake (hunk 5 above)
Comment (winsoft666) [reply ports/zoe/portfile.cmake]: Thanks, I have update cmake.

File: ports/zoe/portfile.cmake [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+        -DZOE_BUILD_SHARED_LIBS:BOOL=${ZOE_BUILD_SHARED_LIBS}
+        -DZOE_USE_STATIC_CRT:BOOL=${ZOE_USE_STATIC_CRT}
+        -DZOE_BUILD_TESTS:BOOL=OFF
+)
+
+vcpkg_cmake_install()
+
+vcpkg_cmake_config_fixup(CONFIG_PATH share/zoe)
+
+file(INSTALL "${CMAKE_CURRENT_LIST_DIR}/usage" DESTINATION "${CURRENT_PACKAGES_DIR}/share/${PORT}")
```
Comment (FrankXie05) [inline ports/zoe/portfile.cmake:24]:
```suggestion
```
No need to add the `usage`, cmake will automatically export the correct usage.

**Review by FrankXie05 — CHANGES_REQUESTED** [review]

File: ports/zoe/portfile.cmake (hunk 6 above)
Comment (winsoft666) [reply ports/zoe/portfile.cmake]: > No need to add the `usage`, cmake will automatically export the correct usage.

Updated

**Review by FrankXie05 — APPROVED** [review]

**Pull request merged**